initial attempt at compression support

daplho opened the pull request:
Thanks for contributing to Logstash! If you haven't already signed our CLA, here's a handy link: https://www.elastic.co/contributor-agreement/

markwalkom commented:
@daplho can you please sign the CLA. I will ask someone to review the code.

daplho commented:
Signed.

daplho commented:
Anything going on here @markwalkom? I'm definitely willing to collaborate with whomever is having a look fyi.

markwalkom commented:
I've asked one of the Logstash team to take a look at this :)

daplho commented:
awesome, have them talk to me directly or via this ticket

jordansissel commented:
Code looks good. Can you add some tests?

daplho commented:
hey @jordansissel, not all is great unfortunately, here is some information I had included in the original 'Issue' I created for this:

"Unfortunately, Manticore (and the underlying Apache CloseableHttpClient) are not playing nice and are creating an gzipped body with an invalid header. Anybody here have some expertise in the area? I've tried lots of combinations of the options that http_client_builder and request_builder expose without any success so far."

daplho commented:
@jordansissel did you get a chance to have a look at this? do you need more detail around the issue I'm describing?

jordansissel commented:
I haven't looked further due to time constraints l, but I understand what
you are saying and I don't think we need more detail. Thank you :)

On Wed, Jul 19, 2017 at 7:07 PM daplho <notifications@github.com> wrote:

> @jordansissel <https://github.com/jordansissel> did you get a chance to
> have a look at this? do you need more detail around the issue I'm
> describing?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/logstash-plugins/logstash-output-http/pull/66#issuecomment-316575298>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6vrVFbe_0Pnfw5TnL1SxuPSlzuyiks5sPrZRgaJpZM4OR2m7>
> .
>

daplho commented:
Cool, are you well-versed with Manticore?

jordansissel commented:
I have some experience with manticore, but I'm not sure what help I'll be this week as I have only just returned from time-off and have ~10 other PRs on my todo list that I am working on ;\

daplho commented:
@jordansissel Checking in with you to see if you're getting close to this one and if I can provide some useful information or discussion as you work toward making sure it works with Manticore as expected?

AnthonyClink commented:
Any community support I could offer for this issue? This feature would aid in a base requirement we have.

daplho commented:
Hello Antony, please feel free to test out my changes. As I mentioned earlier, they are not quite working as I have limited understanding of Manticore but they are a good starting point.

ph commented:
@daplho I am taking at this PR, let's bring that feature to completion!

daplho commented:
Hey @ph, excited to have someone work on this, it's a valuable feature! Please read over this entire discussion - I've indicated above that while my MR does implement compression, it is not working correctly. Also, I'd be more than happy to give you any more info you might need.

ph commented:
@daplho thanks, I will dig deeper and provide some guidelines :)

ph commented:
@daplho Okay, I know what is wrong with the current implementation, you were really close. I've cooked a simple HTTP endpoint with sinatra and a rack decompression handler.

When I try your code without any changes, on the remote end I get a 500 errors complaining about an invalidate gzip stream (or header?). By doing a bit more investigation we can see that the compressed body is generated as **UTF-8**, when the server receive it, it think its **<Encoding:ASCII-8BIT>**. 

You can resolve that issue by making sure the _StringIO_ use binary encoding.

```ruby
  # gzip data
  def gzip(data)
    gz = StringIO.new
    gz.set_encoding("BINARY") # this fix the issue.
    z = Zlib::GzipWriter.new(gz)
    z.write(data)
    z.close
    gz.string
  end
```

By doing this, I can see the correct data sent from logstash on my custom http endpoint.

```
REQUEST BODY: {"@version":"1","host":"sashimi.local","@timestamp":"2017-08-22T19:30:52.318Z","message":"ohollaaaa"}
```

With that changes and added tests we will be able to merge that new feature, I've made available my custom endpoint at https://gist.github.com/ph/8d00dd05a90609cf83f25758b9839f3d

We can use the rack handler (CompressedRequest) as is in the rspec suite to test for compressed data.

ph commented on lib/logstash/outputs/http.rb:
```diff
@@ -305,6 +315,15 @@ def event_body(event)
 
+  # gzip data
+  def gzip(data)
+    gz = StringIO.new
```
We need to make sure we use binary encoding on the StringIO.
```
gz.set_encoding("BINARY") # this fix the issue.
```

daplho replied:
Making it happen right now ;)

ph commented on lib/logstash/outputs/http.rb:
```diff
@@ -80,6 +81,9 @@ class LogStash::Outputs::Http < LogStash::Outputs::Base
   config :format, :validate => ["json", "form", "message"], :default => "json"
 
+  # Set this to true if you want to enable gzip compression for your message
+  config :compression, :validate => :boolean
```
Can we rename this option to `http_compression` this will make sure we the same naming convention as our elasticsearch output, see the doc at https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-http_compression

ph replied:
We also need to update the static doc at https://github.com/logstash-plugins/logstash-output-http/blob/master/docs/index.asciidoc with that new option so it will appear on our website.

daplho commented:
@ph, I've added the "BINARY" fix, updated the option to `http_compression` and updated the docs. I believe the last thing would be adding tests, would you be able to assist with those?

ph commented:
@daplho I can a test suite for it, I am investigating a hang with the suite, is doesn't happen all the time, so I think something is odd with the async behavior and how the test suite is structured.

daplho commented:
Thanks @ph. Can you commit into my fork and branch or do I need to give you special permissions?

ph commented:
@daplho I've pushed a PR at #72 with a commit that include some tests, I've forked your PR and I basically just took the rack handler example and what I've used with sinatra and ported that to the test suite.

I also keep attribution, Everything seems to run fine.

Just to be complete, I have also tested this feature with apache2 and _mod_deflate_ using the **SetInputFilter** directive and it work, example config:

```
Listen 7005

<VirtualHost *:7005>
   SetInputFilter DEFLATE
    ProxyPreserveHost On
    ProxyPass / http://0.0.0.0:4567/
    ProxyPassReverse / http://0.0.0.0:4567/

    ServerName localhost
</VirtualHost>
```

So I will merge your PR, merge the tests PR bump the version + changelog.

ph commented:
closing this since #72 is merged and contains this code.

ph commented:
@daplho the plugins are published 4.4.0 and 5.1.0

daplho commented:
@ph one more thing...have you tried installing the 5.1.0 version of the plugin? It appears that the `logstash-mixin-http_client` is causing some versioning grief:

ph commented:
@dalpho I am running into the same error here, everything works fine with
5.4.2+, I will do a bit of investigation, I am not sure why we bump the
mixin http client and made it incompatible.

On Thu, Aug 24, 2017 at 2:50 PM daplho <notifications@github.com> wrote:

> @ph <https://github.com/ph> one more thing...have you tried installing
> the 5.1.0 version of the plugin? It appears that the
> logstash-mixin-http_client is causing some versioning grief:
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/logstash-plugins/logstash-output-http/pull/66#issuecomment-324723923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAACgOi6NpmhhG-PpkLM8BDQYgieszvNks5sbcXdgaJpZM4OR2m7>
> .
>
-- 
ph

ph commented:
I believe we had a problem with a backward compatibility issue, we have removed an option in the mixin, but by doing that we force the plugins to use a newer version.

I do have a workaround, to make it work.

```
 bin/logstash-plugin uninstall logstash-input-http_poller
Successfully removed logstash-input-http_poller
 bin/logstash-plugin uninstall logstash-output-http
Successfully removed logstash-output-http
bin/logstash-plugin install --version 4.4.0 logstash-output-http
Validating logstash-output-http-4.4.0
Installing logstash-output-http
Installation successful
bin/logstash-plugin list --verbose logstash-output-http
logstash-output-http (4.4.0)

daplho commented:
I've also identified an alternative by using an older version of the `logstash-mixin-http_client`. Specifically, if I adjust the gemfile by replacing

```
s.add_runtime_dependency "logstash-mixin-http_client", ">= 6.0.0", "< 7.0.0"
```

with

```
s.add_runtime_dependency "logstash-mixin-http_client", ">= 5.0.0", "< 6.0.0"
```

and rebuild the plugin then it install successfully.

I tend to like your approach because it uses the latest mixin but I do appreciate mine because there will be no hoops to jump through once we rebuild the plugin with the older version. What do you think?

ph commented:
@daplho The problem is changing the requirements, could break users configuration, I need to put some more thoughts in it.

daplho commented:
@ph well, if we only do it for versions 4.4.0 and 5.1.0 which were not working already it shouldn't matter. Am I missing something?

daplho commented:
@ph did you get a chance to decide on the best way to move forward on this?

ph commented:
@daplho I've created https://github.com/logstash-plugins/logstash-output-http/pull/76 to start a discussion to relax them.

daplho commented:
Thanks @ph - I will keep my eye on [#76](https://github.com/logstash-plugins/logstash-output-http/pull/76).
